### PR TITLE
Position from velocity

### LINKED ISSUE #153: uv2latlon: integrates horizontal velocity to give latitude and longitude. See jLab function of the same name
to be called `position_from_velocity`

### PR DESCRIPTION
**Pull request by milancurcic**
* [x] Basic functionality
* [x] Can specify time axis
* [x] N-dimensional
* [x] Examples
* [x] Tests

Closes #153

### CODE REVIEW

**Comment by milancurcic** [comment]
Though a lot of these bells and whistles are yet to be implemented, you can already try it. I can run `velocity_from_position` on a GDP trajectory to get (u, v) and then use that on `position_from_velocity` to get almost the same trajectory back. We should be able to get exactly the same trajectory back, i.e. a round-trip (lon, lat) -> (u, v) -> (lon, lat) works, so that will be one of the tests that need to pass.

**Review by selipot — CHANGES_REQUESTED** [review]

File: clouddrift/analysis.py [hunk 1]
```diff
@@ -371,6 +371,89 @@ def segment(
...
+    difference_scheme : str, optional
+        The difference scheme to use for computing the position. Can be
+        "forward" or "backward". Default is "forward".
+    time_axis : int, optional
+        The axis of the time array. Default is -1.
+
+    Returns
+    -------
+    x : np.ndarray
+        A two-dimensional array of eastings or longitudes.
```
[the remark this answers is not in the document]
Comment (milancurcic) [reply clouddrift/analysis.py:413]: I think so but we should use whatever terminology our target audience understands most easily.

File: clouddrift/haversine.py [hunk 2]
```diff
@@ -109,3 +110,53 @@ def bearing(
...
+    longitude in degrees, distance in meters, and bearing in radians, based on
+    the spherical law of cosines.
+
+    The formula is:
+
+    φ2 = asin( sin φ1 ⋅ cos δ + cos φ1 ⋅ sin δ ⋅ cos θ )
+    λ2 = λ1 + atan2( sin θ ⋅ sin δ ⋅ cos φ1, cos δ − sin φ1 ⋅ sin φ2 )
+
+    where (φ, λ) is (lat, lon) and θ is bearing, all in radians.
+    Bearing is defined as zero toward East and increasing counter-clockwise.
```
[the remark this answers is not in the document]
Comment (milancurcic) [reply clouddrift/haversine.py:128]: I truly mean increasing here :)

File: clouddrift/analysis.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (selipot) [reply clouddrift/analysis.py:413]: Being the first audience I say let's use _displacements_

File: clouddrift/haversine.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (selipot) [reply clouddrift/haversine.py:128]: increasing counterclockwise is correct but a bearing is an angle not a rotation rate. As such, it is either positive or negative with respect to East. I think _positive counterclockwise_ is therefore the correct wording.

[the remark this answers is not in the document]
Comment (milancurcic) [reply clouddrift/haversine.py:128]: I don't follow this. A change of an angle is a rotation, and clockwise is a rotation. An angle can't be clockwise (whether positive or negative) because it's not a rotation, but it can increase clockwise, because an increase of an angle is a rotation. That said, I'll implement your wording of choice.

[the remark this answers is not in the document]
Comment (selipot) [reply clouddrift/haversine.py:128]: Ok we're debating semantics, which is fun. But yes, my preference is _positive counterclockwise_.

[the remark this answers is not in the document]
Comment (milancurcic) [reply clouddrift/haversine.py:128]: Let's keep these suggestions unresolved until they're implemented, otherwise, they get hidden.

[the remark this answers is not in the document]
Comment (selipot) [reply clouddrift/haversine.py:128]: Sorry, can't seem to do anything right ;)

**Comment by milancurcic** [comment]
Unlike forward or backward, the centered scheme is not supposed to round-trip. This test is ill-posed.

Forward:

$$
x_{n+1} := x_n + u_n \Delta t
$$

$$
x_{n+1} := x_n + \dfrac{x_{n+1} - x_n}{\Delta t} \Delta t
$$

$$
x_{n+1} := x_{n+1}
$$

Centered:

$$
x_{n+1} := x_n + u_{n+\dfrac{1}{2}} \Delta t
$$

$$
x_{n+1} := x_n + \dfrac{1}{2} (u_n + u_{n+1}) \Delta t
$$

$$
x_{n+1} := x_n + \dfrac{1}{2} \dfrac{x_{n+1} - x_{n-1}  + x_{n+1} - x_{n}}{2\Delta t} \Delta t
$$

$$
x_{n+1} := -\dfrac{1}{4} x_{n-1} + \dfrac{3}{4} x_{n} + \dfrac{1}{4} x_{n+1} + \dfrac{1}{4} x_{n+2}
$$

In other words, since the centered difference scheme smooths $2\Delta x$ waves and the centered integration scheme averages over two points, so the round-trip using the centered scheme smooths over 4 points. We should thus not expect round-trip to machine precision. I'll update the test to a reduced tolerance.

**Comment by selipot** [comment]
Great.

File: clouddrift/sphere.py [hunk 3]
```diff
@@ -43,9 +43,6 @@ def recast_lon(lon: np.ndarray, lon0: Optional[float] = -180) -> np.ndarray:
     --------
     :func:`recast_lon360`, :func:`recast_lon180`
     """
-    if np.isscalar(lon):
-        lon = np.array([lon])
-
```
Comment (milancurcic) [inline clouddrift/sphere.py:48]: @philippemiron do you know why we haven't been accepting a scalar in `recast_lon`? When testing `haversine.position_from_distance_and_bearing`, I came to want to pass a scalar to this. All the sphere tests pass with the change in this commit, and I can't tell that any aspect of the function changed for the worse. Is there some gotcha here that I don't see?

**Comment by milancurcic** [comment]
@selipot @philippemiron what's your opinion about the range of longitudes emitted by `haversine.position_from_distance_and_bearing`? If you start from lat1, lon1 = 0, 179, and want to go 2 degrees to the East, without recasting the result is lat2, lon2 = 0, 181. However,  I like my longitudes to default to [-180, 180] range, so I applied `recast_lon(lon, -180)` to the result of  `haversine.position_from_distance_and_bearing`. Do you like this behavior, or should the haversine function not do any recasting for the user?

I'm torn because I like the [-180, 180] range for longitudes, and at the same time, good functions should do one thing.

**Comment by milancurcic** [comment]
> I'm torn because I like the [-180, 180] range for longitudes, and at the same time, good functions should do one thing.

I read what I wrote and am now preferring more to remove the recasting. My personal preference shouldn't come above a principle of a good design.

**Comment by philippemiron** [comment]
I understand the problem but I think that just avoiding doing anything is better. That way you get [2, 182] and you can recast and similar if you are using ]0, 360] range.

File: clouddrift/analysis.py [hunk 4]
```diff
@@ -371,6 +371,207 @@ def segment(
...
+def position_from_velocity(
+    u: np.ndarray,
+    v: np.ndarray,
+    time: np.ndarray,
+    x_origin: float,
+    y_origin: float,
+    coord_system: Optional[str] = "spherical",
+    integration_scheme: Optional[str] = "forward",
+    time_axis: Optional[int] = -1,
+) -> Tuple[xr.DataArray, xr.DataArray]:
```
Comment (milancurcic) [inline clouddrift/analysis.py:383]:
```suggestion
) -> Tuple[np.ndarray, np.ndarray]:
```

File: clouddrift/analysis.py [hunk 5]
```diff
@@ -371,6 +371,207 @@ def segment(
...
+    >>> v = np.array([1., 1., 1., 1.])
+    >>> time = np.array([0., 1., 2., 3.]) * 1e5
+    >>> x, y = position_from_velocity(u, v, time, 0, 0)
+    >>> x
+    array([0.        , 0.89839411, 2.69584476, 5.39367518])
+    >>> y
+    array([0.        , 0.89828369, 1.79601515, 2.69201609])
+
+    Integrating across the antimeridian and recasting longitude to the
+    [-180, 180] range:
```
Comment (milancurcic) [inline clouddrift/analysis.py:481]:
```suggestion
    Integrating across the antimeridian (dateline) by default does not
    recast the resulting longitude:
```

File: clouddrift/analysis.py [hunk 6]
```diff
@@ -371,6 +371,207 @@ def segment(
...
+    y : np.ndarray
+        An array of meridional displacements or latitudes.
+
+    Examples
+    --------
+
+    Simple integration on a plane, using the forward scheme by default:
+
+    >>> import numpy as np
+    >>> from clouddrift.analisys import position_from_velocity
```
Comment (milancurcic) [inline clouddrift/analysis.py:451]:
```suggestion
    >>> from clouddrift.analysis import position_from_velocity
```

File: clouddrift/analysis.py [hunk 7]
```diff
@@ -371,6 +371,207 @@ def segment(
...
+
+    dt = np.diff(time_)
+
+    if integration_scheme == "forward":
+        x[..., 1:] = np.cumsum(u_[..., :-1] * dt, axis=-1)
+        y[..., 1:] = np.cumsum(v_[..., :-1] * dt, axis=-1)
+    elif integration_scheme == "backward":
+        x[..., 1:] = np.cumsum(u_[1:] * dt, axis=-1)
+        y[..., 1:] = np.cumsum(v_[1:] * dt, axis=-1)
+    elif integration_scheme == "centered":
```
Comment (milancurcic) [inline clouddrift/analysis.py:545]:
```suggestion
    if integration_scheme.lower() == "forward":
        x[..., 1:] = np.cumsum(u_[..., :-1] * dt, axis=-1)
        y[..., 1:] = np.cumsum(v_[..., :-1] * dt, axis=-1)
    elif integration_scheme.lower() == "backward":
        x[..., 1:] = np.cumsum(u_[1:] * dt, axis=-1)
        y[..., 1:] = np.cumsum(v_[1:] * dt, axis=-1)
    elif integration_scheme.lower() == "centered":
```

**Review by selipot — APPROVED** [review]

File: clouddrift/analysis.py [hunk 8]
```diff
@@ -371,6 +371,207 @@ def segment(
...
+        y[..., 1:] = np.cumsum(0.5 * (v_[..., :-1] + v_[..., 1:]) * dt, axis=-1)
+    else:
+        raise ValueError(
+            'integration_scheme must be "forward", "backward", or "centered".'
+        )
+
+    if coord_system == "cartesian":
+        x += x_origin
+        y += y_origin
+    elif coord_system == "spherical":
```
Comment (milancurcic) [inline clouddrift/analysis.py:556]:
```suggestion
    if coord_system.lower() == "cartesian":
        x += x_origin
        y += y_origin
    elif coord_system.lower() == "spherical":
```
